### ISSUE TITLE
Why adding "ON DELETE SET NULL" on foreign keys?

### ISSUE THREAD

**Issue by AlexanderMatveev** [ent/ent#1374]

**Comment by AlexanderMatveev** [comment]
Accidentally posted without description.
Here are my models:

```golang
func (User) Edges() []ent.Edge {
	return []ent.Edge{
		//...
		edge.To("streams", Stream.Type),
		/...
	}
}

func (Stream) Edges() []ent.Edge {
	return []ent.Edge{
		// ...
		edge.From("owner", User.Type).
			Ref("streams").
			Unique().
			Required(),
		// ...
	}
}
```

Migrations provided:
```sql
-- ...
ALTER TABLE "streams" ADD CONSTRAINT "streams_users_streams" FOREIGN KEY("user_streams") REFERENCES "users"("id") ON DELETE SET NULL;
-- ...
```

Why `ON DELETE SET NULL` is added `Required()` in edge?

**Comment by a8m** [comment]
Hello @AlexanderMatveev and thanks for opening this issue.

> Why ON DELETE SET NULL is added `Required()` in edge?

The `ON DELETE SET NULL` clause is added by default to all non-M2M relations because their columns are not configured as `NOT NULL` columns in the database (edge checks happen only on the application level). However, you can configure the foreign-key reference action differently to one of [these actions](https://github.com/ent/ent/blob/master/dialect/entsql/annotation.go#L130-L136) using [edge annotations](https://entgo.io/docs/schema-annotations/#foreign-keys-configuration). 

Does it solve your problem?

**Comment by AlexanderMatveev** [comment]
Hi @a8m, thanks for response.

My point is that default PostgreSQL action is `NO ACTION`, which same as `RESTRICT`, see `referential_action` here https://www.postgresql.org/docs/current/sql-createtable.html. And I think it must be a strong argument of making `SET NULL` as default, because this creates an unexpected integrity violation.

And, about `OnDelete` annotation. There is no `OnDelete` field in 0.7.0 release: https://github.com/ent/ent/blob/v0.7.0/dialect/entsql/annotation.go

**Comment by a8m** [comment]
> And, about OnDelete annotation. There is no OnDelete field in 0.7.0 release: https://github.com/ent/ent/blob/v0.7.0/dialect/entsql/annotation.go

You can pull the version in master. It's stable.

> My point is that default PostgreSQL action is NO ACTION, which same as RESTRICT, see referential_action here https://www.postgresql.org/docs/current/sql-createtable.html. And I think it must be a strong argument of making SET NULL as default, because this creates an unexpected integrity violation.

I agree with this argument. The design (and the implementation) was started based on our needs, and in some cases, we didn't use foreign-keys at all. I'm fine with starting a proposal for changing (or extending via option) the schema migration behavior for required-edges that hold a foreign-key to set the database column to `NOT NULL`, and change the referential action to `NO ACTION`.

That means if a parent row needs to be deleted, the child row needs to be deleted or updated with a different reference.

**Comment by JoleMile** [comment]
Any work being done on this? Being able to set foreign-keys to `NOT NULL` seems extremely useful.

**Comment by a8m** [comment]
> Being able to set foreign-keys to `NOT NULL` seems extremely useful.

This is already the default behavior. Can you explain what are you trying to do and it's not working?

**Comment by a8m** [comment]
Hey @AlexanderMatveev and @JoleMile 👋
Sorry for the delay, but the issue was resolved with #1703 - Thanks to [github.com/atlas/ariga](https://github.com/atlas/ariga). That means if you're using MySQL and need to migrate existing columns, you should use the [`WithAtlas`](https://entgo.io/docs/migrate#atlas-integration) option.

cc @rotemtam

**Comment by Liooo** [comment]
@a8m

> My point is that default PostgreSQL action is NO ACTION, which same as RESTRICT, see referential_action here https://www.postgresql.org/docs/current/sql-createtable.html. And I think it must be a strong argument of making SET NULL as default, because this creates an unexpected integrity violation.

Does this not apply to non-required edges as well?

https://github.com/ent/ent/blob/0d1257f5f21ae64d7f24decd64967a3042efc06c/entc/gen/graph.go#L822-L824
